### DISCUSSION TITLE
Making EditorInteract public

### DISCUSSION THREAD

**Discussion by Jdance-Media** [SmartlyDressedGames/Unturned-3.x-Community#4374] [Ideas]
It seems that several fields in EditorInteract were made public, I assume with the intention of it being usable to outside libraries. It would be logical for us to be able to access worldHit, objectHit, and logicHit without needing to use reflection, as you seem to have the class already properly setup for us to use it.

**Comment by SDGNelson** [comment]
Unfortunately, most old `public` members should really have been `internal`. When I first started work on Unturned 3 I'd never written code that would be used be other developers before, so it wasn't on my radar whatsoever. (All?) new code should be safe in that regard, but even several years into Rocket first launching I wasn't giving it enough attention.

**Reply by Jdance-Media** [reply]
Understandable in that case. The main problem I have with EditorInteract is the terrain tool. Right now it is incredibly confusing to modify the heightmap tool settings, as some of the fields/properties are in two different classes, `TerrainEditor` and `DevkitLandscapeToolHeightmapOptions`. Brush strength is a public nonstatic member within TerrainEditor but heightmap mode is static. And then the rest of the settings are in the options class. Moving over those two settings still left in TerrainEditor for consistency would spare us from using reflection. Recreating the raycast hits is easier said than done, so I suppose only true problem with EditorInteract being nonpublic is the TerrainEditor.

**Reply by SDGNelson** [reply]
It looks like if the goal is to modify the settings everything is currently public. There are some helper properties to modify the *current* tool's strength, but it just points to the public fields in `DevkitLandscapeToolHeightmapOptions` like `flattenStrength` or `smoothStrength`.

**Reply by Jdance-Media** [reply]
Hmm you are correct, but `heightmapBrushStrength` still requires me to have an instance of `TerrainEditor`, which is stored in `EditorInteract`, which is internal. Since these are helper properties that do not access any instance members, it would be preferable to make them static. I'm not sure what you had planned with them being virtual though. 

I'll close this now knowing that I can simply recreate what the `heightmapBrushStrength` property is doing and avoid reflection.

**Reply by DanielWillett** [reply]
It would be cool to see some other way to mark code as 'probably shouldn't be used but could still be temporarily useful', since there are a lot of times when we do need to use internal code to do more special case stuff, and reflection can be a bit slower and make code a lot harder to manage and read. Entity Framework (very popular microsoft library for database models) uses XML comments for this (see below). Also, it kinda seems like almost all client-side code is being marked as internal now.

It can be kinda frustrating making entire wrapper classes and dynamic methods for pretty much all new client-side code (enter [UIAccessTools.cs](https://github.com/DanielWillett/DevkitServer/blob/master/API/UI/UIAccessTools.cs) from stage left).

![image](https://github.com/SmartlyDressedGames/Unturned-3.x-Community/assets/12886600/d0c20ac6-1b02-40bf-9498-eb162ae34316)

**Reply by sunnamed434** [reply]
If you guys really want to use smth and without reflection you can use one nuget package, it makes all things public, see [https://github.com/EvolutionPlugins/Dummy/blob/openmod/Dummy%2FDummy.csproj#L37-L45](here)
